tools/cldr-code/src/main/java/org/unicode/cldr/util/personname/SimpleNameObject.java

```diff
@@ -1,85 +1,140 @@
+ * Immutable
+ */
+public class SimpleNameObject implements NameObject {
+    private final ULocale nameLocale;
+    private final Map<Field, Map<Set<Modifier>, String>> patternData;
 
-        @Override
-        public Set<Field> getAvailableFields() {
-            Set<Field> result = EnumSet.noneOf(Field.class);
-            for (Entry<ModifiedField, String> entry : patternData.entrySet()) {
-                result.add(entry.getKey().getField());
-            }
-            return ImmutableSet.copyOf(result);
+    @Override
+    public Set<Field> getAvailableFields() {
+        return patternData.keySet();
+    }
+
+    /**
+     * Return the best fit.
+     *  The data is organized by field, and then by modifier sets.
+     *  These are in longest-first order, so we take the first one that is not disjoint.
+     *  The ordering is lexicographic among items with the same number of elements, so we favor earlier values in Modifier.values();
+     * We are guaranteed by construction that the last one is empty
+     */
+    @Override
+    public String getBestValue(ModifiedField modifiedField, Set<Modifier> remainingModifers) {
+        // First check for match to field
+        Map<Set<Modifier>, String> fieldData = patternData.get(modifiedField.getField());
+        if (fieldData == null) {
+            return null;
         }
+        Set<Modifier> modifiers = modifiedField.getModifiers();
 
-        @Override
-        public String getBestValue(ModifiedField modifiedField, Set<Modifier> remainingModifers) {
-            // TODO Alex flesh out the SimpleNameObject to handle modifiers
-            // and return the ones that need handling by the formatter
 
-            // For now just return the item.
-            return patternData.get(modifiedField);
+        for (Entry<Set<Modifier>, String> entry : fieldData.entrySet()) {
+            Set<Modifier> set = entry.getKey();
+            if (set.isEmpty() || !Collections.disjoint(remainingModifers, modifiers)) {
+                remainingModifers.addAll(modifiers);
+                remainingModifers.removeAll(set);
+                return entry.getValue();
+            }
         }
+        throw new IllegalArgumentException("Internal error: Every field must have a completely modified value " + fieldData);
+    }
 
-        @Override
-        public ULocale getNameLocale() {
-            return nameLocale;
-        }
+    @Override
+    public ULocale getNameLocale() {
+        return nameLocale;
+    }
 
-        public SimpleNameObject(ULocale nameLocale, ImmutableMap<ModifiedField, String> patternData) {
-            this.nameLocale = nameLocale == null ? ULocale.ROOT : nameLocale;
-            this.patternData = patternData;
+    public SimpleNameObject(ULocale nameLocale, Map<ModifiedField, String> patternData) {
+        this.nameLocale = nameLocale == null ? ULocale.ROOT : nameLocale;
+        Map<Field, Map<Set<Modifier>, String>> _patternData = new EnumMap<>(Field.class);
```

Comment:
BTW, I liked the way this code worked out; by changing the data structure I was able to minimize the number of lookups. One to get the field data, then walk through the cases where the NameObject also had explicit modifier strings. (For our current data, the bulk have just 1 item.)